test/read_write_laspoints.jl

```diff
@@ -0,0 +1,196 @@
+get_constructor_values(::Type{LasPoint4}, t::TestLasPoint) = [get_constructor_values(LasPoint1, t)..., t.wave_packet_descriptor_index, t.wave_packet_byte_offset, t.wave_packet_size_in_bytes, t.wave_return_location, t.wave_x_t, t.wave_y_t, t.wave_z_t]
+get_constructor_values(::Type{LasPoint5}, t::TestLasPoint) = [get_constructor_values(LasPoint3, t)..., t.wave_packet_descriptor_index, t.wave_packet_byte_offset, t.wave_packet_size_in_bytes, t.wave_return_location, t.wave_x_t, t.wave_y_t, t.wave_z_t]
+
+# point formats 6 - 10
+get_constructor_values(::Type{LasPoint6}, t::TestLasPoint) = [t.x, t.y, t.z, t.intensity, t.flag_byte_1, t.flag_byte_2, t.classification, t.user_data, t.extended_scan_angle, t.pt_src_id, t.gps_time]
+get_constructor_values(::Type{LasPoint7}, t::TestLasPoint) = [get_constructor_values(LasPoint6, t)..., t.red, t.green, t.blue]
+get_constructor_values(::Type{LasPoint8}, t::TestLasPoint) = [get_constructor_values(LasPoint7, t)..., t.nir]
+get_constructor_values(::Type{LasPoint9}, t::TestLasPoint) = [get_constructor_values(LasPoint6, t)..., t.wave_packet_descriptor_index, t.wave_packet_byte_offset, t.wave_packet_size_in_bytes, t.wave_return_location, t.wave_x_t, t.wave_y_t, t.wave_z_t]
+get_constructor_values(::Type{LasPoint10}, t::TestLasPoint) = [get_constructor_values(LasPoint8, t)..., t.wave_packet_descriptor_index, t.wave_packet_byte_offset, t.wave_packet_size_in_bytes, t.wave_return_location, t.wave_x_t, t.wave_y_t, t.wave_z_t]
+
+
+const SUPPORTED_LAS_POINT_FORMATS = [
+    LasPoint0,
+    LasPoint1,
+    LasPoint2,
+    LasPoint3,
+    LasPoint4, # with waveform
+    LasPoint5, # with waveform
+    LasPoint6,
+    LasPoint7,
+    LasPoint8,
+    LasPoint9, # with waveform
+    LasPoint10, # with waveform
+]
+
+@testset "Test different LAS Versions" begin
+    workdir = mktempdir()
+    for version_minor = 1 : 4
+        for point_type_index = 1 : length(SUPPORTED_LAS_POINT_FORMATS)
+            point_type = SUPPORTED_LAS_POINT_FORMATS[point_type_index]
+            number_of_points = 10
+            max_return_number = version_minor > 3 ? 15 : 5
+            return_numbers = rand(1 : max_return_number, max_return_number)
+            return_count = [UInt32(sum(i .== return_numbers)) for i = 1 : 5]
+
+            # use LasPoint1
+            points = [point_type(get_constructor_values(point_type, TestLasPoint())...) for i = 1 : number_of_points]
+            data_record_length = size(point_type)
+
+            point_return_count = version_minor == 4 ? fill(UInt64(0), 15) : fill(UInt32(0), 5)
+            return_number_array = clamp.([return_number(p) for p in points], 1, length(point_return_count))
+            @inbounds for i in 1:length(point_return_count)
+                point_return_count[i] = sum(return_number_array .== i)
+            end
+
+            scale = [0.001, 0.001, 0.001]
+            offset = [0.0, 0.0, 0.0]
```

Comment:
Always use non-zero values in tests, you may miss issues (e.g. not taking offset into account, or XYZ-copy-paste-mistakes) as those are not caught using zeros. `offset = [1.0, 2.0, 3.0]`